flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java

```diff
@@ -205,32 +215,70 @@ public void writeBuffers(List<BufferWithChannel> bufferWithChannels) throws IOEx
             return;
         }
 
-        long expectedBytes = 0;
+        long expectedBytes;
         ByteBuffer[] bufferWithHeaders = new ByteBuffer[2 * bufferWithChannels.size()];
 
+        if (isBroadCastRegion) {
+            expectedBytes = writeBroadcastBuffers(bufferWithChannels, bufferWithHeaders);
+        } else {
+            expectedBytes = writeUnicastBuffers(bufferWithChannels, bufferWithHeaders);
+        }
+
+        BufferReaderWriterUtil.writeBuffers(dataFileChannel, expectedBytes, bufferWithHeaders);
+    }
+
+    private long writeUnicastBuffers(
+            List<BufferWithChannel> bufferWithChannels, ByteBuffer[] bufferWithHeaders) {
+        long expectedBytes = 0;
         for (int i = 0; i < bufferWithChannels.size(); i++) {
-            BufferWithChannel bufferWithChannel = bufferWithChannels.get(i);
-            Buffer buffer = bufferWithChannel.getBuffer();
-            int subpartitionIndex = bufferWithChannel.getChannelIndex();
-            if (subpartitionIndex != currentSubpartition) {
+            int subpartition = bufferWithChannels.get(i).getChannelIndex();
+            if (subpartition != currentSubpartition) {
                 checkState(
-                        subpartitionBuffers[subpartitionIndex] == 0,
+                        subpartitionBuffers[subpartition] == 0,
                         "Must write data of the same channel together.");
-                subpartitionOffsets[subpartitionIndex] = totalBytesWritten;
-                currentSubpartition = subpartitionIndex;
+                subpartitionOffsets[subpartition] = totalBytesWritten;
+                currentSubpartition = subpartition;
             }
 
-            ByteBuffer header = BufferReaderWriterUtil.allocatedHeaderBuffer();
-            BufferReaderWriterUtil.setByteChannelBufferHeader(buffer, header);
-            bufferWithHeaders[2 * i] = header;
-            bufferWithHeaders[2 * i + 1] = buffer.getNioBufferReadable();
+            Buffer buffer = bufferWithChannels.get(i).getBuffer();
+            int numBytes = setBufferWithHeader(buffer, bufferWithHeaders, 2 * i);
+            expectedBytes += numBytes;
+            totalBytesWritten += numBytes;
+            ++subpartitionBuffers[subpartition];
+        }
+        return expectedBytes;
+    }
+
+    private long writeBroadcastBuffers(
+            List<BufferWithChannel> bufferWithChannels, ByteBuffer[] bufferWithHeaders) {
+        if (subpartitionBuffers[0] == 0) {
```

Comment:
Maybe we could add some comment here to explain we only need to set the offset on first call ?